Add support for xbox/windows store games

idbrii opened the pull request:
This builds on top of changes from #8, so that commit is included here.

Uses a method similar to BrianLima/UWPHook to discover app packages. See
https://github.com/BrianLima/UWPHook/blob/master/UWPHook/Resources/GetAUMIDScript.ps1

However, instead of creating a wrapper like UWPHook, we find the exe and
add it directly to steam. This might not work for UWP games, but
probably most games are win32.

I included the `--skip-xbox-library` argument just in case this adds something someone doesn't want. All the other libraries can be skipped by setting their path to somewhere invalid, but xbox doesn't have paths, so this seemed appropriate.

Test
* Genesis Noir
* Monster Train
* Prey (2017)
* Rain on Your Parade
* Spiritfarer
* Supraland
* Tetris Effect

jaydenmilne reviewed (commented):
I tried this out with Halo 5 Forge and Roblox (I don't own any Xbox games for Windows). Halo was detected and it created a shortcut, but it can't launch (the exe icon has a padlock, so I assume UWP weirdness). Roblox is disqualified because it's an `App` and it doesn't have any libs to detect, and glancing at the manifest also appears to be a UWP app.

I'm wondering if there is a way to detect a game is UWP and disqualify it, since it isn't great UX to have add a game and get cryptic error messages from Steam when you try and launch it. We might need to label this feature as 'experimental' or 'beta' or something.

jaydenmilne commented:
If we can figure out if an app is a UWP app, then it seems like we have the pieces to go through [this process](https://www.addictivetips.com/windows-tips/open-uwp-apps-from-command-line-windows-10/) and make a shortcut to `C:\Windows\explorer.exe` with launch options `shell:appsFolder\Microsoft.Halo5Forge_8wekyb3d8bbwe!Ausar` to launch the game. We don't get steam input, but streaming works. 

Just thinking out loud here

jaydenmilne commented:
Final thoughts for merge:
- Fix extra argument
- I think we need to 'disqualify' UWP apps somehow. If we can't figure that out, this needs to be clearly identified as experimental (I had a 0/2 success rate haha) (I can take care of this in the README/etc)
- If we can identify UWP apps, we can try doing the explorer.exe trick to half-effort it.

idbrii commented on steamsync-library/src/xbox.py:
```diff
@@ -0,0 +1,133 @@
+
+    print("Scanning Xbox library...")
+    # Run string instead of scriptfile to circumvent powershell ExecutionPolicy
+    # ("cannot be loaded because running scripts is disabled on this system").
```
Putting it in a string does work, but when debugging I found it easier to have an actual powershell script to run and edit with syntax highlighting.

If it doesn't automatically get included in the package by being in the folder, then putting it as a `'''` string is fine.

idbrii replied:
I did `pip install -e steamsync\steamsync-library` to install from my git checkout and running `steamsync.py --source xbox` correctly detected my xbox games, so I don't think we need to stuff the ps1 script into a python string.

idbrii commented:
I just found that Spiritfarer doesn't actually work (my testing was that games get added, but not that they were launching properly). So I need to run through some additional testing.

The shell:appsFolder trick doesn't work for it either:

    explorer.exe shell:appsFolder\44905ThunderLotusGames.40196900AE792!App

Maybe it can only get executed with whatever trick UWPHook is using.

I'd guess that neither Halo 5 Forge nor Roblox would get detected on the 7be671d version and it's the later commits that try to be more wide-reaching that pick them up.

Looks like the Spiritfarer exe isn't a readable file, so that may be a good way to detect if it will be launchable.

idbrii commented:
I was wrong about Spiritfarer, you can launch via explorer.exe but I had the aumid wrong. Figured out how to get the right aumid and I've changed the code to launch via explorer.exe anything without a MicrosoftGame.config (which seems to indicate it's not a UWP app). I feel a lot better about using the explorer trick for starting games that don't seem to follow modern conventions. Thanks for the suggestion!

Can you try it with Halo 5? It should either no longer be detected or now launch properly.

I tested with a bunch of games. I tried using a Dual Sense gamepad to see if launching from steam would have better support, but no dice.

| detected | runs | dual sense  | game                      |
|----------|------|-------------|---------------------------|
| y        | y    | y           | Carto                     |
| y        | y    | n           | Genesis Noir for Windows  |
| y        | y    | y           | Ghost of a Tale PC        |
| y        | y    | n (partial) | Monster Train             |
| y        | y    | n           | Prey                      |
| y        | y    | n           | Rain on Your Parade       |
| **no**        | --    | --           | Roblox                    |
| y        | y    | n           | Spiritfarer               |
| y        | y    | n           | Supraland                 |
| y        | y    | n           | Tetris® Effect: Connected |
| y        | y    | n           | Unto The End              |



I also added sorting the game list (useful when I was comparing the list in steam to the list in the output) and some formatting changes.

jaydenmilne commented:
Halo 5: detected, launch OK via the explorer.exe hack 🎉(halo 5 forge is ftp by the way so you can test with it). No steam input, as expected.

I'll do a final code review; do you have anything else you want to include in this?

jaydenmilne reviewed (approved):
LGTM

jaydenmilne commented:
Thanks for the work on this, turned out great!

idbrii commented:
> I'll do a final code review; do you have anything else you want to include in this?


🎆 Nope. I'm happy with it.
